mailbox/backup/src/main/java/org/apache/james/mailbox/backup/DefaultMailboxBackup.java

```diff
@@ -0,0 +1,137 @@
+        Stream<MessageResult> messages = allMessagesForUser(accountContents);
+        return saveToStore(mailboxes, messages);
+    }
+
+    private final MailboxManager mailboxManager;
+    private final ArchiveService archiveService;
+    private final BlobStore store;
+
+    private Function<MailboxPath, Optional<MailAccountContent>> getMailboxWithAnnotationsFromPath(MailboxSession session) {
+        return path -> {
+            try {
+                StoreMessageManager messageManager = (StoreMessageManager) mailboxManager.getMailbox(path, session);
+                Mailbox mailbox = messageManager.getMailboxEntity();
+                List<MailboxAnnotation> annotations = mailboxManager.getAllAnnotations(path, session);
+                MailboxWithAnnotations mailboxWithAnnotations = new MailboxWithAnnotations(mailbox, annotations);
+                Stream<MessageResult> messages = Iterators.toStream(messageManager.getMessages(MessageRange.all(), FetchGroupImpl.FULL_CONTENT, session));
+                return Optional.of(new MailAccountContent(mailboxWithAnnotations, messages));
+            } catch (MailboxException e) {
+                LOGGER.error("Error while fetching Mailbox during backup", e);
+                return Optional.empty();
+            }
+        };
+    }
+
+    private List<MailAccountContent> getAccountContentForUser(MailboxSession session) throws MailboxException {
+        MailboxQuery queryUser = MailboxQuery.builder().username(session.getUser().asString()).build();
+        Stream<MailboxPath> paths = mailboxManager.search(queryUser, session).stream().map(mailboxMetaData -> mailboxMetaData.getPath());
+        List<MailAccountContent> mailboxes = paths
+            .flatMap(getMailboxWithAnnotationsFromPath(session).andThen(OptionalUtils::toStream))
+            .collect(Guavate.toImmutableList());
+
+        return mailboxes;
+    }
+
+    private Publisher<BlobId> saveToStore(List<MailboxWithAnnotations> mailboxes, Stream<MessageResult> messages) throws IOException {
+        PipedOutputStream outputStream = new PipedOutputStream();
+        PipedInputStream inputStream = new PipedInputStream();
+        inputStream.connect(outputStream);
+
+        Mono.fromRunnable(Throwing.runnable(() -> archiveService.archive(mailboxes, messages, outputStream)).sneakyThrow())
+            .doOnSuccessOrError(Throwing.biConsumer((result, throwable) -> {
+                if (throwable != null) {
+                    LOGGER.error("An error happened when archiving  messages", throwable);
+                }
+                outputStream.flush();
+                outputStream.close();
+            }))
+            .subscribeOn(Schedulers.elastic())
+            .then().subscribe();
+
+        return store.save(inputStream).map(b -> BlobId.fromString(b.asString()))
+            .doOnSuccessOrError(Throwing.biConsumer((blobId, e) -> inputStream.close()));
```

Comment:
oh I see now why you wanted to have only one blobId concept. But here I guess you want a blobstore blobId, not a mailbox blobId, no?

Reply:
Thanks, i've refactored it to only use the blobstore, blobId. It's much cleaner.